contracts/data-storage/andromeda-graph/src/contract.rs

```diff
@@ -0,0 +1,395 @@
+                    .deps
+                    .querier
+                    .query_wasm_smart(address.clone(), &PointQueryMsg::GetDataOwner {})?;
+                let user: AndrAddr = user_res.owner;
+                let user_addr = user.get_raw_address(&ctx.deps.as_ref())?;
+
+                user_point_coordinate.validate()?;
+
+                USER_COORDINATE.save(ctx.deps.storage, user_addr, &user_point_coordinate)?;
+            } else {
+                return Err(ContractError::InvalidADOType {
+                    msg: Some(format!("ADO Type must be point: {:?}", ado_type)),
+                });
+            }
+        } else {
+            // Not a contract
+            return Err(ContractError::InvalidAddress {});
+        }
+    }
+    Ok(Response::new().add_attributes(vec![attr("action", action), attr("sender", sender)]))
+}
+
+pub fn execute_delete_user_coordinate(
+    ctx: ExecuteContext,
+    user: AndrAddr,
+    action: String,
+) -> Result<Response, ContractError> {
+    let sender = ctx.info.sender;
+    ensure!(
+        ADOContract::default().is_owner_or_operator(ctx.deps.storage, sender.as_ref())?,
+        ContractError::Unauthorized {}
+    );
+    let user_addr = user.get_raw_address(&ctx.deps.as_ref())?;
+
+    USER_COORDINATE.remove(ctx.deps.storage, user_addr);
+
+    Ok(Response::new().add_attributes(vec![attr("action", action), attr("sender", sender)]))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, env: Env, msg: QueryMsg) -> Result<Binary, ContractError> {
+    match msg {
+        QueryMsg::GetMapInfo {} => encode_binary(&get_map_info(deps.storage)?),
+        QueryMsg::GetMaxPointNumber {} => encode_binary(&get_max_point_number(deps.storage)?),
+        QueryMsg::GetAllPoints {} => encode_binary(&get_all_points(deps.storage)?),
+        QueryMsg::GetUserCoordinate { user } => encode_binary(&get_user_coordinate(deps, user)?),
+        _ => ADOContract::default().query(deps, env, msg),
+    }
+}
+
+pub fn get_map_info(storage: &dyn Storage) -> Result<GetMapInfoResponse, ContractError> {
+    let map_info = MAP_INFO
+        .load(storage)
+        .map_err(|_| ContractError::InvalidParameter {
+            error: Some("Map not found".to_string()),
+        });
+    match map_info {
+        Ok(map_info) => Ok(GetMapInfoResponse { map_info }),
+        Err(err) => Err(err),
+    }
+}
+
+pub fn get_max_point_number(
+    storage: &dyn Storage,
+) -> Result<GetMaxPointNumberResponse, ContractError> {
+    let max_point_number = TOTAL_POINTS_NUMBER.load(storage)?;
+    Ok(GetMaxPointNumberResponse { max_point_number })
+}
+
+pub fn get_all_points(storage: &dyn Storage) -> Result<GetAllPointsResponse, ContractError> {
+    let max_point_number = TOTAL_POINTS_NUMBER.load(storage)?;
+
+    let mut res: Vec<(CoordinateInfo, StoredDate)> = Vec::new();
+
+    for point in 1..=max_point_number {
+        let coordinate = MAP_POINT_INFO.load(storage, &point)?;
```

Comment:
This can error if a point is removed. Maybe use `may_load` and only append `Option` values that have a `Some` value.

Reply:
This is also unpaginated so has a potential DOS vector if too many points are added.

Reply:
Using a paginated `range` value may actually be a better approach.

Reply:
Resolved like this:
```
pub fn get_all_points(
    storage: &dyn Storage,
    start: Option<u128>,
    limit: Option<u32>,
) -> Result<GetAllPointsResponse, ContractError> {
    let max_point_number = TOTAL_POINTS_NUMBER.load(storage)?;

    // Set default values for pagination
    let start_point = start.unwrap_or(1); // Start from 1 if no start provided
    let limit = limit.unwrap_or(100); // Default limit to 100 points

    let mut res: Vec<(CoordinateInfo, StoredDate)> = Vec::new();

    // Iterate with pagination
    for point in start_point..=max_point_number {
        if res.len() >= limit as usize {
            break; // Stop when limit is reached
        }

        // Use `may_load` to handle cases where the point may not exist
        if let Some(coordinate) = MAP_POINT_INFO.may_load(storage, &point)? {
            res.push(coordinate);
        }
    }

    Ok(GetAllPointsResponse { points: res })
}
```